### PR TITLE
Switch mmlu serving to be external the same as we do with mt_bench

### PR DESCRIPTION
**Pull request by danmcp**
Like with mt_bench, the new logic supports vllm and llama-cpp and passes the base url of the started instance to the mmlu library.

This enables:
- Consistency with how we are serving from a support perspective
- Support for larger models with sharding across gpus
- Multi-gpu support
- Potential for faster performance with higher serving throughput

Related: 
https://github.com/instructlab/eval/issues/50
https://github.com/instructlab/eval/issues/68

Corresponding Eval change:
https://github.com/instructlab/eval/pull/99

Resolves: #1792

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the
  [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [ ] [Changelog](https://github.com/instructlab/instructlab/blob/main/CHANGELOG.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

**Comment by danmcp** [comment]
Note: Tests will fail until https://github.com/instructlab/eval/pull/99 is released

File: src/instructlab/model/evaluate.py [hunk 1]
```diff
@@ -529,6 +522,8 @@ def evaluate(
         merge_system_user_message,
     )
 
+    validate_model(model)
```
Comment (nathan-weinberg) [inline src/instructlab/model/evaluate.py:525]: Could this go before we try to initialize the evaluator above?

Comment (danmcp) [reply src/instructlab/model/evaluate.py:525]: My logic was get_evaluator checks whether the model param was set.  So it made sense to check whether it was valid after that step.

Let me try again to move the validate into get_evaluator.  That's probably the best solution other than splitting get_evaluator up into a two different functions.

Comment (danmcp) [reply src/instructlab/model/evaluate.py]: Updated to move everything into get_evaluator.  As I did the work I remembered the reason the other reason I put it was because it was a lot more code to put it into get_evaluator and to rework the tests.  But it is a better change now I think.  I also added validate for judge models and base models which was missing before.

**Comment by nathan-weinberg** [comment]
Eval PR has been merged

**Review by nathan-weinberg — APPROVED** [review]
Looked over everything - code makes sense to me - thanks for resolving this @danmcp!

File: src/instructlab/model/evaluate.py [hunk 2]
```diff
@@ -187,6 +163,32 @@ def get_evaluator(
...
+                fg="red",
+            )
+            raise click.exceptions.Exit(1)
+        click.secho(
+            f"Using local model found at '{model_path}' for '{model_arg}'",
+            fg="blue",
+        )
+    else:
+        click.secho(
+            f"Model could not be found at '{model}'",
```
Comment (alimaredia) [inline src/instructlab/model/evaluate.py:186]: Could you change this to `Model could not be found at '{model]' for '{model_args}'.

Comment (danmcp) [reply src/instructlab/model/evaluate.py]: Changed, thanks!

File: src/instructlab/model/evaluate.py [hunk 3]
```diff
@@ -187,6 +163,32 @@ def get_evaluator(
         )
 
 
+def validate_model(model: str, model_arg: str = "--model"):
```
Comment (leseb) [inline src/instructlab/model/evaluate.py:166]: I feel like this is a generic function that could benefit other parts of the code as well. How about moving it to utils? Any code that uses "model" could use that right?

**Comment by alimaredia** [comment]
@danmcp are we going to need to bump the minimum version of `instructlab-eval` in requirements.txt since external serving with MMLU can't be done with versions before `0.2.1`?

**Comment by danmcp** [comment]
> @danmcp are we going to need to bump the minimum version of `instructlab-eval` in requirements.txt since external serving with MMLU can't be done with versions before `0.2.1`?

Done

File: src/instructlab/model/evaluate.py (hunk 3 above)
Comment (danmcp) [reply src/instructlab/model/evaluate.py:166]: The first bits seem pretty generic.  I don't think we can move all the sechos and exception raising though.  I could have this wrong, but I would also think we want to avoid backend calls in utils.

That would leave:

```
        valid_model = False
        if model_path.is_dir():
            valid_model = backends.is_model_safetensors(model_path)
        elif model_path.is_file():
            valid_model = backends.is_model_gguf(model_path)
```

Which could be made a util method.  Maybe that would be a useful method in backend?

Comment (leseb) [reply src/instructlab/model/evaluate.py:166]: I guess as a backend function would work too. As for the `click.secho` we could replace with `print` and `click.exceptions` we could use a generic exception. We don't have to do this now though, we can revisit it later if other portions of the code need it.

File: src/instructlab/model/evaluate.py [hunk 4]
```diff
@@ -691,7 +695,21 @@ def evaluate(
...
 
         elif benchmark == Benchmark.MMLU:
-            overall_score, individual_scores = evaluator.run()
+            server = None
+            try:
+                server, api_base = launch_server(
+                    ctx,
+                    model,
+                    model,
+                    None,
```
Comment (leseb) [inline src/instructlab/model/evaluate.py:704]: Why is `max_workers` set to None for MMLU?

Comment (danmcp) [reply src/instructlab/model/evaluate.py:704]: max_workers is only for the mt_bench variants

Comment (leseb) [reply src/instructlab/model/evaluate.py:704]: Ok but given that MMLU now uses the same serving backend and that both backend supports max_workers AFAIK I don't understand why max_workers wouldn't be applicable. Could you please clarify for me? Thanks!

File: src/instructlab/model/evaluate.py (hunk 3 above)
Comment (danmcp) [reply src/instructlab/model/evaluate.py:166]: Most of my concern over the exits/exceptions/prints is about where the logic exists to exit and return messages to the user.  Generally when I think of a util method as not making choices about what gets showed to a user or if a program should exit.  But rather it would return indicators (exceptions, booleans, etc.) to higher level code that would then make those choices.

So if we did want to move all the logic to backend or utils.  I would expect the util function to raise specific exceptions for each of the error conditions, those exceptions would then be handled by higher level code and specific messages printed before exiting.

All of that is doable but I am not sure it's going to be any less verbose.  Moving the valid_model code block from above to backend is a pretty easy win though if we want to do that.

Happy to be corrected if I am thinking about this the wrong way.

File: src/instructlab/model/evaluate.py (hunk 4 above)
Comment (danmcp) [reply src/instructlab/model/evaluate.py:704]: max_workers isn't used by the serving, it's used by the evaluation logic itself.  Specifically, mt_bench uses a pool of workers to run through the evaluation questions and max_workers is the tuning param for that pool.

File: src/instructlab/model/evaluate.py (hunk 3 above)
Comment (leseb) [reply src/instructlab/model/evaluate.py:166]: It's all good, we can keep it here and I agree with you logic around "utils" too. Thanks.

**Review by leseb — APPROVED** [review]
